--- model_structure_process.py.bak	2023-12-28 23:14:28.280176800 +0000
+++ model_structure_process.py	2023-12-29 00:25:19.100176800 +0000
@@ -61,12 +61,17 @@ class StructureProcess:
                                qkv_list,
                                proj_list,
                                mhsa_ln_list):
-        expected_matmul_lang = 2
-        if len(matmul_list) != expected_matmul_lang:
-            return
+        matmul_list = matmul_list[:2]

         proj_mat, qkv_mat = matmul_list
         if proj_mat.node.out_features * 3 == qkv_mat.node.out_features:
+            input_nodes = [in_node for in_node in qkv_mat.input_nodes]
+            qkv_mat_list = [
+                item for item in input_nodes[0].output_nodes
+                if item.op_type == ModuleType.LINEAR.value
+            ]
+            if len(qkv_mat_list) != 1:
+                return
             qkv_list.append([qkv_mat.name_in_network])
             proj_list.append(proj_mat.name_in_network)
             mhsa_ln_list.append(ln_list[0].name_in_network)
@@ -75,7 +80,7 @@ class StructureProcess:
                 return
             input_nodes = [in_node for in_node in qkv_mat.input_nodes]
             qkv_mat_list = [item for item in input_nodes[0].output_nodes if item.op_type == ModuleType.LINEAR.value]
-            if len(qkv_mat_list) != 3:
+            if len(qkv_mat_list) not in [2,3]:
                 return

             qkv_list.append([item.name_in_network for item in qkv_mat_list])